### ISSUE TITLE
Cannot export API with our Internal trusted certificates from API Manager.

### ISSUE THREAD

**Issue by stevebower59** [Axway-API-Management-Plus/apimanager-swagger-promote#157]
I have been having an issue exporting APIs which have Trusted certificates signed by our internal CA.   I've been able to successfully export APIs which have external CAs with no issue so far.
The APIs work fine through the gateway.
  
When exporting the API using api-export I get the following traceback.
```
0      INFO  nvironmentProperties| Trying to load environment properties from file: env.properties ... not found.
1575   INFO     APIManagerAdapter| API-Manager version is: 7.6.2 SP3
1940   INFO     APIManagerAdapter| Found existing API on path: '/dhp/onbase/api/v2/document' (published) (ID: 'a00fe368-dd72-46b2-bda5-0d63dbeb8c63')
6645   INFO  IExportConfigAdapter| Going to export API into folder: ..\apis\dhp-onbase-api-v2-document
6845   ERROR            ExportApp| Can't write API-Configuration file.
com.axway.apim.lib.AppException: Can't write API-Configuration file.
        at com.axway.apim.api.export.APIExportConfigAdapter.saveAPILocally(APIExportConfigAdapter.java:119)
        at com.axway.apim.api.export.APIExportConfigAdapter.exportAPIs(APIExportConfigAdapter.java:59)
        at com.axway.apim.ExportApp.run(ExportApp.java:124)
        at com.axway.apim.ExportApp.main(ExportApp.java:35)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: (was java.lang.NullPointerException) (through reference chain: com.axway.apim.api.export.ExportAPI["caCerts"]->java.util.ArrayList[0]->com.axway.apim.swagger.api.properties.cacerts.CaCert["certFile"])
        at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:394)
        at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:353)
        at com.fasterxml.jackson.databind.ser.std.StdSerializer.wrapAndThrow(StdSerializer.java:316)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFieldsFiltered(BeanSerializerBase.java:779)
        at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:153)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:119)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:79)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:18)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:727)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:719)
        at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:155)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
        at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue(ObjectMapper.java:3906)
        at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3160)
        at com.axway.apim.api.export.APIExportConfigAdapter.saveAPILocally(APIExportConfigAdapter.java:117)
        ... 3 more
Caused by: java.lang.NullPointerException
        at com.axway.apim.swagger.api.properties.cacerts.CaCert.getCertFile(CaCert.java:171)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:688)
        at com.fasterxml.jackson.databind.ser.impl.SimpleBeanPropertyFilter.serializeAsField(SimpleBeanPropertyFilter.java:208)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFieldsFiltered(BeanSerializerBase.java:771)
        ... 15 more
```

**Comment by cwiechmann** [comment]
Hi @stevebower59, I see the reason for the NPE. In order to improve/stabilize this it would be good if you could share the certificates making trouble here.
If this is not possible, print the certificate info here:
`openssl x509 -in certificate.crt -text -noout`

**Comment by stevebower59** [comment]
[badcerts.zip](https://github.com/Axway-API-Management-Plus/apimanager-swagger-promote/files/3643223/badcerts.zip)

Hi @cwiechmann,
I've attached 3 of the certs that were causing trouble, along with the CA, and Intermediate certs in the chain.  Hope this helps.

**Comment by cwiechmann** [comment]
I have attached all 5 certificates to my API and was still able to export it. I made some further tests and realized that you have overwritten the auto-generated name of the certificate in the API-Manager UI. 

From something like this:
CN=National Oilwell Varco Enterprise CA-HOU,DC=nov,DC=com
To maybe:
National Oilwell Varco Enterprise CA-HOU Root CA

I made the mistake to use the Certificate-Name in order to generate the resulting filename. This wasn't the best choice, as it could be overwritten/changed. 

I'm going to change this code and instead use the Certificate-Alias which makes much more sense.

**Comment by cwiechmann** [comment]
Hi @stevebower59, with this SNAPSHOT-Version you should be to download all API. It would be great if you can confirm. 

Thanks!
Chris

[apimanager-swagger-promote-1.6.3-SNAPSHOT.zip](https://github.com/Axway-API-Management-Plus/apimanager-swagger-promote/files/3647865/apimanager-swagger-promote-1.6.3-SNAPSHOT.zip)

**Comment by stevebower59** [comment]
Hi @cwiechmann,

I tried on three of my API's and it looks like I'm getting a very similar  NPE stacktrace for the api-export for the 1.6.3-SNAPSHOT version that I just downloaded this morning.
  
Stacktrace follows:

C:\Axway\swagger-promote-1.6.3-SNAPSHOT\scripts>.\api-export -h catalogdev.nov.com -port 443 -u apiadmin -p s0a0agdev21db2019 -l ..\apis -a /dhp/onbase/api/v2/uploaddocument
------------------------------------------------------------------------
API-Manager Promote: 1.6.3-SNAPSHOT - E X P O R T

To report issues or get help, please visit:
https://github.com/Axway-API-Management-Plus/apimanager-swagger-promote
------------------------------------------------------------------------

0      INFO  nvironmentProperties| Trying to load environment properties from file: env.properties ... not found.
1575   INFO     APIManagerAdapter| API-Manager version is: 7.6.2 SP3
1875   INFO     APIManagerAdapter| Found existing API on path: '/dhp/onbase/api/v2/uploaddocument' (published) (ID: '777984a6-d90a-439a-9b5f-fcc0280a4bad')
6510   INFO  IExportConfigAdapter| Going to export API into folder: ..\apis\dhp-onbase-api-v2-uploaddocument
6710   ERROR            ExportApp| Can't write API-Configuration file.
com.axway.apim.lib.AppException: Can't write API-Configuration file.
        at com.axway.apim.api.export.APIExportConfigAdapter.saveAPILocally(APIExportConfigAdapter.java:119)
        at com.axway.apim.api.export.APIExportConfigAdapter.exportAPIs(APIExportConfigAdapter.java:59)
        at com.axway.apim.ExportApp.run(ExportApp.java:124)
        at com.axway.apim.ExportApp.main(ExportApp.java:35)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: (was java.lang.NullPointerException) (through reference chain: com.axway.apim.api.export.ExportAPI["caCerts"]->java.util.ArrayList[0]->com.axway.apim.swagger.api.properties.cacerts.CaCert["certFile"])
        at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:394)
        at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:353)
        at com.fasterxml.jackson.databind.ser.std.StdSerializer.wrapAndThrow(StdSerializer.java:316)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFieldsFiltered(BeanSerializerBase.java:779)
        at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:153)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:119)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:79)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:18)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:727)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:719)
        at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:155)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
        at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue(ObjectMapper.java:3906)
        at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3160)
        at com.axway.apim.api.export.APIExportConfigAdapter.saveAPILocally(APIExportConfigAdapter.java:117)
        ... 3 more
Caused by: java.lang.NullPointerException
        at com.axway.apim.swagger.api.properties.cacerts.CaCert.getCertFile(CaCert.java:171)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:688)
        at com.fasterxml.jackson.databind.ser.impl.SimpleBeanPropertyFilter.serializeAsField(SimpleBeanPropertyFilter.java:208)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFieldsFiltered(BeanSerializerBase.java:771)
        ... 15 more

**Comment by stevebower59** [comment]
Would it help for me to attach the api-config.json (incomplete because of the Exception), and the Swagger definition file?

**Comment by stevebower59** [comment]
Swagger and api-config.json created using api-export utility - version 1.6.3-SNAPSHOT.

[api-config.zip](https://github.com/Axway-API-Management-Plus/apimanager-swagger-promote/files/3648426/api-config.zip)

**Comment by cwiechmann** [comment]
Hmm, that's strange. @stevebower59, can you please re-run the same with the attached version. It logs more information about the certificate.
[apimanager-swagger-promote-1.6.3-SNAPSHOT.zip](https://github.com/Axway-API-Management-Plus/apimanager-swagger-promote/files/3648747/apimanager-swagger-promote-1.6.3-SNAPSHOT.zip)

I need the line starting with the below to understand what happens:
`"CaCert [certFile=.....`

**Comment by stevebower59** [comment]
Thanks @cwiechmann,

I retried with the new api-export, and the line starting with the CaCert debug is as follows:

68321  INFO                CaCert| CaCert [certFile=null, name=EMAILADDRESS=barry.dolton@nov.com, CN=srvgdywbjob02.nov.com, OU=Wellbore Technologies Information Systems, O=National Oilwell Varco, L=Houston, ST=TX, C=US, alias=EMAILADDRESS=barry.dolton@nov.com, CN=srvgdywbjob02.nov.com, OU=Wellbore Technologies Information Systems, O=National Oilwell Varco, L=Houston, ST=TX, C=US, subject=EMAILADDRESS=barry.dolton@nov.com, CN=srvgdywbjob02.nov.com, OU=Wellbore Technologies Information Systems, O=National Oilwell Varco, L=Houston, ST=TX, C=US, version=3, notValidBefore=1537884007000, notValidAfter=1632492007000, signatureAlgorithm=RSA (2048 bits), sha1Fingerprint=35:73:9F:E4:90:CC:5E:7F:EA:08:67:ED:2A:2F:08:49:44:5E:29:25, md5Fingerprint=44:0B:7E:7D:4E:82:C5:73:FA:2C:95:BC:EC:DF:FA:A0, expired=false, notYetValid=false]

It still failed in the same way.

**Comment by cwiechmann** [comment]
@stevebower59, thanks for sharing the Log-Information. Based on that, I have identified the root-cause and fixed it now.

**Comment by cwiechmann** [comment]
Fix delivered with release 1.6.3
